lib/requester.js

```diff
@@ -160,7 +162,7 @@ function startRequest(operation) {
       throw new Error('request for unknown authenticator type '+authType);
     }
   } else {
-    if(operation.client.connectionParams.apiKey && !operation.accessToken){
+    if(operation.client.connectionParams.apiKey && (!operation.accessToken || operation.expiration <= (new Date()))){
```

Comment:
What happens if there's a race condition and the accessToken expires in the time between line 165 and when the server receives the request?

Reply:
Right, I think there are windows where this won't be reliable. 

The approach in the Java Client is to not bother with the expiration time and just assume that the first 403 it receives is due to the token expiring. It then generates a new token and tries the request again. If that also results in a 403, then it propagates the 403 to the user.

Here's the code for checking on a 403 - https://github.com/marklogic/java-client-api/blob/develop/marklogic-client-api/src/main/java/com/marklogic/client/impl/okhttp/MarkLogicCloudAuthenticationConfigurer.java#L160 .

And then assuming that multiple threads can be using the same client instance (I assume that's possible in the DMSDK impl in the Node Client), you should synchronize the code block that generates the new token as shown here - https://github.com/marklogic/java-client-api/blob/develop/marklogic-client-api/src/main/java/com/marklogic/client/impl/okhttp/MarkLogicCloudAuthenticationConfigurer.java#L183 . That accounts for 2 or more threads determining that a new token is needed at the same time - we only want 1 of those threads to "win" and generate a new token, and then the other threads will realize the token has been updated and will use the new token.

Reply:
Let me update the errorListener as well